Allow HLS video streaming, and use fallback video sources as download URL

pomegranited opened the pull request:
### Description

Allows HLS videos to be streamed.

And because HLS videos cannot be downloaded using our current video download library, then fallback video URLs are used for downloading.  

The following sources are used to locate a valid download URL:

* available encodings, when the Video Pipeline is enabled.
* HTML5 sources, when the Video Pipeline is disabled.

**JIRA**

[OSPR-2019](https://openedx.atlassian.net/browse/OSPR-2019)

### Notes

Functionality depends on API changes in https://github.com/edx/edx-platform/pull/16590, but the app continues to work even if this change is not deployed yet.

Builds on https://github.com/edx/edx-app-ios/pull/1014, which differentiates between streamable and downloadable video URLs.

We'd like to get this merged before the next edx-app-ios release.

### How to test this PR

Ensure that the changes from https://github.com/edx/edx-platform/pull/16590 are running in your LMS, and import this [video_test_course.tar.gz](https://github.com/edx/edx-app-ios/files/1493756/video_test_course.Dl2Kjh.tar.gz), which contains permutations of all the various video streaming/download scenarios.

Alternately, run against this sandbox server, using:
```yaml
API_HOST_URL: 'https://pr16590.sandbox.opencraft.hosting'
OAUTH_CLIENT_ID: '14cdc120d61349b7eeba'
```

1. Run the iOS app.
1. Login as a student to the test course.
1. Visit each of the course subsections in turn, noting the expected Streaming and Download behaviours for each unit below.
    *note* Once you've downloaded a video, that video will be used in the player, regardless of what the original streaming video URL is.  So to test the permutations below, try streaming first, then download.

* YouTube:
    * Standard: YouTube 
        * Streaming: YouTube only; will not play on iOS app, but will play in browser.
        * Download: none
    * Standard w Fallback:  
        * Streaming: 0.05 mp4 (same as MP4 - 1), different from YouTube video
        * Download: yes, same as stream
    * HLS live stream:
        * Streaming: YouTube only; will not play on iOS app, but will play in browser.
        * Download: none
    * HLS w fallback:
        * Streaming: 0.05 mp4 (same as MP4 - 1)
            *Note* The main video is configured to use the same YouTube HLS used in the previous unit, but `USING_VIDEO_PIPELINE: false` causes the fallback URL to be streamed instead.
        * Download: Same as stream.
* MP4 
    * MP4 - 1
        * Streaming: 0.05 mp4 
        * Download: yes, same as stream
    * MP4 - 1 w fallback
        * Streaming: 0.05 mp4 
        * Download: yes, same as stream
            *Note* Fallback video is configured to be the 0:27 mp4 used by MP4 - 2, but because the main video 
            can be downloaded, the fallback video is ignored.
    * MP4 - 2
        * Streaming: 0:27 mp4 
        * Download: yes, same as stream
    * MP4 - 2 w fallback
        * Streaming: 0:27 mp4 
        * Download: yes, same as stream
            *note* Fallback video is configured to be the 0:05 mp4 used by MP4 - 1, but because the main video 
            can be downloaded, the fallback video is ignored.
* HLS .m3u8
    * HLS (10 min):
        * Streaming: 10:35 video
        * Download: none
    * HLS (10 min) w fallback:
        * Streaming: 10:35 video
        * Download: 0:27 mp4, same as MP4 - 2
    * Vimeo
        * Streaming: 8:22 Cloudera m3mu
        * Download: none
    * Vimeo w fallback
        * Streaming: 8:22 Cloudera m3mu
        * Download: 8:22 Cloudera mp4

### Screenshots

`USING_VIDEO_PIPELINE: false`
* Course page
 ![using_video_pipeline false course](https://user-images.githubusercontent.com/7556571/34403760-2b854134-ebfa-11e7-9324-a73992440929.png)
* YouTube section
  ![using_video_pipeline false youtube](https://user-images.githubusercontent.com/7556571/34403766-36f68cee-ebfa-11e7-99e4-e367c751925b.png)
* MP4 section
 ![using_video_pipeline false mp4](https://user-images.githubusercontent.com/7556571/34403769-3c3932ba-ebfa-11e7-9435-5278300f3838.png)
* HLS section
   ![using_video_pipeline false hls](https://user-images.githubusercontent.com/7556571/34403774-4256a6dc-ebfa-11e7-85f4-63206a75a581.png)

`USING_VIDEO_PIPELINE: true`
* Course page
  ![using_video_pipeline true course](https://user-images.githubusercontent.com/7556571/34403793-5a471b32-ebfa-11e7-9439-b61f955e0338.png)
* YouTube section
  ![using_video_pipeline true youtube](https://user-images.githubusercontent.com/7556571/34403797-5edcb454-ebfa-11e7-86a1-9d64e13c957f.png)
* MP4 section
  ![using_video_pipeline true mp4](https://user-images.githubusercontent.com/7556571/34403800-63d4d25c-ebfa-11e7-89d7-9e0301e92e0a.png)
* HLS section
   ![using_video_pipeline true hls](https://user-images.githubusercontent.com/7556571/34403805-692a1384-ebfa-11e7-87b8-be31a68c1ab0.png)

### Reviewers

- [x] @bradenmacdonald 
- [ ] edX TBD

edx-webhook commented:
Thanks for the pull request, @pomegranited! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=1033&repo=edx%2Fedx-app-ios"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-2019](https://openedx.atlassian.net/browse/OSPR-2019) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-app-ios/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

pomegranited commented:
@saeedbashir cf [your reply](https://github.com/edx/edx-app-ios/pull/1014#issuecomment-342743152) from https://github.com/edx/edx-app-ios/pull/1014:
I was unable to complete the conversion to using AVFoundation to allow HLS downloads. 
 AVFoundation uses AVAssets, which obscure the downloaded file locations, and so our existing video resource management logic and the video player would need to be re-worked.

So instead, I offer this change, which allows clients who are not using the video processing service (VAL) to provide downloadable fallback video URLs in their courses.

nedbat commented:
@edx/mobile-product Please take a look.
@edx/ios-devs This will need engineering approval, thanks.

saeedbashir commented:
@nedbat We have some other high priority tasks on our plates. Will get back to it soon.

pomegranited commented:
Thanks for the update @saeedbashir !

salman2013 commented on Source/OEXVideoSummary.m:
```diff
@@ -88,6 +89,20 @@ - (id)initWithDictionary:(NSDictionary*)dictionary {
+
+        // If the preferred encoding video URL is downloadable, then use that
+        self.downloadURL = nil;
+        if ([OEXVideoSummary isDownloadableVideoURL:self.videoURL]) {
```
The following case will be miss 
if fallback is not supported or there is no encoding and the videoURL is of extension .m3u8 it would not play any video.
If there is just .m3u8 extension videos in "all_sources" it would not play any video.

It should play the .video with m3u8 extension video in the above case.

pomegranited replied:
I don't understand?  I do not change the `self.videoURL`, which is what's used [when playing or streaming videos](https://github.com/open-craft/edx-app-ios/blob/cbff31879113eefa96888fd375b71aa14fd2fc9e/Source/OEXVideoPlayerInterface.m#L89).

But I add an additional `downloadURL`, which is used when deciding whether a video can be downloaded.  That URL is judged for downloadability by its extension.  And since we can't download `.m3u8` videos, then if there is just an .m3u8 extension videos in "all_sources", it would not *download* any video.

salman2013 commented on Source/OEXVideoSummary.m:
```diff
@@ -88,6 +89,20 @@ - (id)initWithDictionary:(NSDictionary*)dictionary {
         if (![[OEXConfig sharedConfig] isUsingVideoPipeline]) {
```
you can test the fallback not supported scenario by enabling the "USING_VIDEO_PIPELINE" flag.
Please try to test the app by enabling the "isUsingVideoPipeline" flag, you can find this in "OEXConfig+AppFeatures.swift" class.

pomegranited replied:
Thanks, this explanation helped.

salman2013 commented:
@pomegranited The download icon are showing with videos on section level which are not downloadable.

pomegranited commented:
@salman2013 
> The download icon are showing with videos on section level which are not downloadable.
Can you give me some more details on this situation?  Are these pipeline videos?  What are the primary and secondary URLs?

salman2013 commented:
Section Level

![img_0011](https://user-images.githubusercontent.com/4454296/33715989-4c9de5cc-db76-11e7-9820-e0ce4763b70b.PNG)

SubSection Level
![img_0013](https://user-images.githubusercontent.com/4454296/33716016-60d260ea-db76-11e7-83ff-340e2e80dff9.PNG)


The download Progress View
![img_0012](https://user-images.githubusercontent.com/4454296/33716031-6c3dc744-db76-11e7-9144-48096c49faec.PNG)

@pomegranited you can see in the attached screenshots on section level how the download icon are showing for videos which are not in supported formats.

For videos encodings edX is using the [Videos pipeline](https://github.com/edx/edx-video-pipeline) and only supporting  mobile_high and mobile_low videos encoding for mobile.
For the open edX community we are providing support for fallback video encoding with flag  "USING_VIDEO_PIPELINE". If the "USING_VIDEO_PIPELINE" is disabled only then fallback can be considered. It would be great if you can check behaviour of your PR by enabling "USING_VIDEO_PIPELINE", because on edX we are using this flag and not supporting fallback encoding anymore.

Now there are two things which are happening
* First is by enabling the pipeline it would not support the .mp4 and fallback extension so if the video url is of .m3u8 extension then the code should support that which is not working.
* Second is when we enable the pipeline flag the download icons are showing at section level which should not because the videos format are not supported.

> What are the primary and secondary URLs?

Can you please elaborate it?

pomegranited commented:
Thank you for your review, @salman2013 !  I don't have any experience with the video pipeline, so please see aef2651 and let me know if I've misinterpreted.

> > What are the primary and secondary URLs?
> Can you please elaborate it?

I did not know if you were using my test course, but your screenshots answered my question.  Thank you!

pomegranited commented on Source/OEXVideoSummary.m:
```diff
@@ -174,6 +189,23 @@ - (BOOL) isSupportedVideo {
+}
+
+- (BOOL) isDownloadableVideo {
+    return (BOOL)self.downloadURL;
```
Using  does not work?  [boolValue](https://developer.apple.com/documentation/foundation/nsstring/1409420-boolvalue) only returns `true` if the string is "Y", "y", "T", "t", or a digit between 1-9.

salman2013 commented on Source/OEXInterface.m:
```diff
@@ -191,8 +191,13 @@ + (BOOL)isURLForVideo:(NSString*)URLString {
+        for (NSString *extension_option in VIDEO_URL_EXTENSION_OPTIONS) {
+        // Check each string in VIDEO_URL_EXTENSION_OPTIONS to see if the URL has a valid file extension for a video
+            if([URLString localizedCaseInsensitiveContainsString:extension_option]) {
+                return YES;
```
this can be further improve like
return [URLString localizedCaseInsensitiveContainsString:extension_option];

pomegranited replied:
@salman2013 This won't work..  We need to loop through all the possible extensions before we can return `No`.

salman2013 commented:
@pomegranited Please remove the conflicts. Thanks

pomegranited commented:
@salman2013 I've rebase on `master` to resolve the conflict, and addressed your suggestion above.  Is it ok for me to squash?  Any more changes required?

Thank you!

salman2013 commented on Source/OEXInterface.m:
```diff
@@ -176,8 +176,13 @@ + (BOOL)isURLForVideo:(NSString*)URLString {
     if([URLString rangeOfString:URL_SUBSTRING_VIDEOS].location != NSNotFound) {
```
We need to remove this check "([URLString rangeOfString:URL_SUBSTRING_VIDEOS].location != NSNotFound) " now as this has deprecated so please remove the if condition and just keep the else part to check for valid url with extension.

The method will look like 

+ (BOOL)isURLForVideo:(NSString*)URLString {
    for (NSString *extension_option in VIDEO_URL_EXTENSION_OPTIONS) {
        // Check each string in VIDEO_URL_EXTENSION_OPTIONS to see if the URL has a valid file extension for a video
        if([URLString localizedCaseInsensitiveContainsString:extension_option]) {
            return YES;
        }
    }
    return NO;
}

pomegranited replied:
Addressed by c79c747

salman2013 commented:
>  I've rebase on master to resolve the conflict, and addressed your suggestion above. Is it ok for me to squash? Any more changes required?

@pomegranited I have add few more comments to make some changes. Thanks

pomegranited commented on README.rst:
```diff
@@ -17,7 +17,7 @@ Building
     git clone https://github.com/edx/edx-app-ios
 
-2. Open ``edX.xcworkspace``.
+2. Open ``edX.xcworkspace`` using XCode 8.3.3
```
Addressed by bf5aeeb

pomegranited commented:
Thanks @salman2013 !  Is it ok for me to squash now?

salman2013 commented:
@saeedbashir Please take a look into PR as it looks fine to me. Thanks

pomegranited commented:
Thanks @salman2013 !

FYI @saeedbashir , I've squashed the commits from c79c747 into 6585855, so this is ready to merge whenever you approve.

saeedbashir commented:
@pomegranited Sure it will also help in reviewing. What I observer, HLS videos are not supported if `USING_VIDEO_PIPELINE` flag is true. We want to support **streamable HLS** videos in all cases.

pomegranited commented:
Thanks @saeedbashir -- I thought I fixed these issues in a previous commit (see https://github.com/edx/edx-app-ios/pull/1033#issuecomment-350722529).

Is it still not working correctly?  Do you mind to send me a course export so I can test on my devstack with the course and videos you're using?

saeedbashir commented:
Yes, unfortunately, it still has some issues. I was using the sandbox that is mentioned in PR description. 
This is what I got after setting the value of `USING_VIDEO_PIPELINE` to true.
![simulator screen shot - iphone 7 - 2017-12-26 at 17 31 35](https://user-images.githubusercontent.com/5606473/34370982-cb0913fe-eaea-11e7-88d2-a1683bfd4238.png)

pomegranited commented:
@saeedbashir Ah ok.. let's see if I understand correctly:

To find a downloadable video URL:
* First, check the [supported encodings](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR93) returned by the Course Blocks API.
* Second, check the [streamable videoURL](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR106).
* Finally, check the [`all_sources` list](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR111), as returned by the Course Blocks API.

So then, the only thing that the `USING_VIDEO_PIPELINE` setting does is change [whether "fallback" is in the a supported encodings list](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR103), which happens only when `USING_VIDEO_PIPELINE: False`.

Does this sound correct?

saeedbashir commented:
Yup and in case of `USING_VIDEO_PIPELINE: true` we will be doing the same but only for streamable videos. Even if some videos have both streamable and download URLs in ` streamable videoURL` or `all_sources` then we will be considering only streamable URL.

saeedbashir commented:
@pomegranited You don't need to squash because we use git squash while merging PR into master.

pomegranited commented:
@saeedbashir 
> Yup and in case of `USING_VIDEO_PIPELINE: true` we will be doing the same but only for streamable videos. Even if some videos have both streamable and download URLs in `streamable videoURL` or `all_sources` then we will be considering only streamable URL.

Ok cool.. 

How do I determine which video URLs are "streamable"?

Is this handled by [not including the "fallback" encoding in the supported encodings list](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR103)?

Or is it handled by the URL extension check we're doing here with [`VIDEO_URL_EXTENSION_OPTIONS`]()https://github.com/edx/edx-app-ios/pull/1033/files#diff-be681ce662f3fff24b6cc423333815e2R24) and [`ONLINE_ONLY_VIDEO_URL_EXTENSIONS `](https://github.com/edx/edx-app-ios/pull/1033/files#diff-be681ce662f3fff24b6cc423333815e2R25)

saeedbashir commented:
We need to include `fallback` encoding [here](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR93) only for the iteration.

> How do I determine which video URLs are "streamable"?

Any URL which has the extension `.m3u8` is a streamable URL.

pomegranited commented:
@saeedbashir 
> Any URL which has the extension .m3u8 is a streamable URL.

This change doesn't actually add support for downloading `.m3m8` videos.

Because `.m3m8` URLs are really playlists, not videos themselves, you need a special handler to allow them to be downloaded.  This change was deemed to be too difficult for us this round, because it required converting to using the AVFoundation library to provide the video downloads as AVAssets, and to provide the streaming player. 

That's why we introduced this additional [`ONLINE_ONLY_VIDEO_URL_EXTENSIONS `](https://github.com/edx/edx-app-ios/pull/1033/files#diff-be681ce662f3fff24b6cc423333815e2R25), which indicates which URLs can only be streamed when online, and cannot be downloaded.

So instead, we're allowing course authors to provide alternate, downloadable video URLs when they're using `.m3m8` videos for streaming, so that at least there's an option for mobile users to download the course videos.

saeedbashir commented:
Sorry I wasn't clear. By streamable it was meant 'ONLINE PLAYABLE' videos. Let me make it easy for you with an example.

Let's consider video 'HLS (10 min) w fallback' it has URLs with .mp4 and .m3u8
A user will be able to play and download the video if `USING_VIDEO_PIPELINE: false`, but with `USING_VIDEO_PIPELINE: true` user will only be able to play it online.
Do let me know if it is still confusing?

pomegranited commented:
@saeedbashir Isn't that what you're seeing in [the screenshot you sent above](https://github.com/edx/edx-app-ios/pull/1033#issuecomment-354051501)?  There's no "download" link shown for 'HLS (10 min) w fallback', because the fallback `.mp4` URL is allowed when the app checks for downloadable videos.  But the HLS video can be streamed, correct?

saeedbashir commented:
@pomegranited This is how it should behave after enabling/disabling `USING_VIDEO_PIPELINE`
![imgonline-com-ua-twotoone-tadlutg1q2](https://user-images.githubusercontent.com/5606473/34374087-e4cdddb8-eb00-11e7-9f4f-4e7d60df39cf.jpg)

pomegranited commented:
Thanks @saeedbashir !  I can see where I was going wrong now.  Thank you for your patience -- plan to have an update for you in the next 24h.

pomegranited commented:
@saeedbashir bb9b3fd fixes the issues with the streaming behaviour for `USING_VIDEO_PIPELINE: true` that you noted in https://github.com/edx/edx-app-ios/pull/1033#issuecomment-354066987.  I've also:
* added tests for the expected `isSupportedVideo` and `isDownloadableVideo` behaviours for the permutations affected here.
* reduced the diff for `OEXVideoSummary initWithDictionary` by moving the [`downloadURL` property to a method](https://github.com/edx/edx-app-ios/pull/1033/files#diff-c882ae0794557ae865d8d59d70a27deeR198) instead of storing a variable.
* added screenshots to the PR description to show the effect of both `USING_VIDEO_PIPELINE` options.

saeedbashir commented:
@pomegranited All the scenarios work perfectly fine. But, I like the previous approach, populating the `downloadURL` in `OEXVideoSummary initWithDictionary`, because with the introduction of getter for `downloadURL` looping and all the required checks being performed every time `downloadURL` being accessed and also it introduces a new instance variable `allSources`.
If the code readability is concerned here then `downloadURL` populating logic can be moved to a method that can take `allSources` or `any other` required data as params. What do you think?

saeedbashir commented:
@pomegranited One more point just for the sake of knowledge, the PR is specifically for the handling of `HLS Videos` then why the title "Use HTML5 source as download URL" was given?

pomegranited commented:
Thank you @saeedbashir !

> If the code readability is concerned here then downloadURL populating logic can be moved to a method that can take allSources or any other required data as params. What do you think?
Yep, that works fine for me. Done with f3948c6b133750255930c50b6d528c7b74ba9a14.

> One more point just for the sake of knowledge, the PR is specifically for the handling of HLS Videos then why the title "Use HTML5 source as download URL" was given?

My original PR built on https://github.com/edx/edx-app-ios/pull/1014 which added the HLS streaming capability, but I neglected to update the title to reflect that this PR replaces that one.  Have updated it, thank you for noticing that!

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-app-ios&entry.867055334=https://github.com/edx/edx-app-ios/pull/1033&entry.1484655318=https://github.com/pomegranited&entry.752974735=2017-11-16+14:31&entry.1917517419=2018-01-02+12:04&entry.2133058324=Yes) so we can improve your experience in the future.

pomegranited commented:
Thank your for your help and patience, @saeedbashir !
